sycl/include/CL/sycl/types.hpp

```diff
@@ -2274,6 +2274,73 @@ __SYCL_DECLARE_FLOAT_VECTOR_CONVERTERS(double)
 #undef __SYCL_DECLARE_SCALAR_BOOL_CONVERTER
 #undef __SYCL_USE_EXT_VECTOR_TYPE__
 
+/// is_device_copyable is a user specializable class template to indicate
+/// that a type T is device copyable. If is_device_copyable is specialized such
+/// that is_device_copyable_v<T> == true on a T that does not satisfy all
+/// the requirements of a device copyable type, the results are unspecified.
+template <typename T, typename = void>
+struct is_device_copyable : std::false_type {};
+
+template <typename T>
+struct is_device_copyable<
+    T, std::enable_if_t<std::is_trivially_copyable<T>::value>>
+    : std::true_type {};
+
+#if __cplusplus >= 201703L
+template <typename T>
+inline constexpr bool is_device_copyable_v = is_device_copyable<T>::value;
+#endif // __cplusplus >= 201703L
+
+// std::tuple<> is implicitly device copyable type.
+template <> struct is_device_copyable<std::tuple<>> : std::true_type {};
+
+// std::tuple<Ts...> is implicitly device copyable type if each type T of Ts...
+// is device copyable.
+template <typename T, typename... Ts>
+struct is_device_copyable<std::tuple<T, Ts...>>
+    : detail::bool_constant<is_device_copyable<T>::value &&
+                            is_device_copyable<std::tuple<Ts...>>::value> {};
+
+namespace detail {
+template <typename T, typename = void>
+struct IsDeprecatedDeviceCopyable : std::false_type {};
+
+// TODO: using C++ attrubute [[deprecated]] or the macro __SYCL2020_DEPRECATED
+// does not produce expected warning message for the type 'T'.
+template <typename T>
+struct __SYCL2020_DEPRECATED("This type isn't device copyable in SYCL 2020")
+    IsDeprecatedDeviceCopyable<
+        T, std::enable_if_t<std::is_trivially_copy_constructible<T>::value &&
+                            std::is_trivially_destructible<T>::value &&
+                            !is_device_copyable<T>::value>> : std::true_type {};
+
+#ifdef __SYCL_DEVICE_ONLY__
+template <typename T, unsigned NumToCheck> struct CheckDeviceCopyableHelper;
+
+template <typename T> struct CheckDeviceCopyableHelper<T, 0> {};
+
+template <typename T> struct CheckDeviceCopyableHelper<T, 1> {
+  using FieldT = decltype(__builtin_field_type(T, 0));
+  static_assert(is_device_copyable<FieldT>::value ||
+                    detail::IsDeprecatedDeviceCopyable<FieldT>::value,
+                "Not device copyable!");
```

Comment:
I wonder if there is a way to print this type and the kernel name or line this kernel is declared as part of the error message.

Reply:
Not a good way to get the name, but any static-assert would at least include its instantiation, which would make someone good at deciphering template-instantiation-traces able to tell the type.

One thing that MIGHT help is to move this static-assert into its own template:

````
template<typename FieldT>
DiagnoseFieldTy {
static_assert(... and so forth);
};
````

The advantage of course is that the type itself appears in the error-instantiation, rather than the field index.

Reply:
I'll replace the message to "Specified type is not device copyable".
@romanovvlad the error message gives user information about what type causes the problems. For example, for an object of class named 'NT' the error message is:
D:\gs_public\llvm\build\install\bin\..\include\sycl\CL/sycl/types.hpp:2333:3: error: static_assert
      failed due to requirement 'is_device_copyable< NT, void >::value || detail::IsDeprecatedDeviceCopyable< NT,
      void >::value' "Not device copyable!"
  static_assert(is_device_copyable< FieldT >::value ||
  ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:\gs_public\llvm\build\install\bin\..\include\sycl\CL/sycl/types.hpp:2331:7: note: in instantiation
      of template class 'sycl::detail::CheckDeviceCopyableHelper< (lambda at types.cpp:64:32), 2 >' requested here
    : CheckDeviceCopyableHelper< T, NumToCheck - 1 > {
      ^
D:\gs_public\llvm\build\install\bin\..\include\sycl\CL/sycl/types.hpp:2331:7: note: in instantiation
      of template class 'sycl::detail::CheckDeviceCopyableHelper<(lambda at types.cpp:64:32), 3>' requested here
D:\gs_public\llvm\build\install\bin\..\include\sycl\CL/sycl/types.hpp:2340:7: note: in instantiation
      of template class 'sycl::detail::CheckDeviceCopyableHelper<(lambda at types.cpp:64:32), 4>' requested here
    : CheckDeviceCopyableHelper<FuncT, __builtin_num_fields(FuncT)> {};
      ^
D:\gs_public\llvm\build\install\bin\..\include\sycl\CL/sycl/handler.hpp:1182:5: note: in
      instantiation of template class 'sycl::detail::CheckDeviceCopyable< (lambda at types.cpp:64:32) >' requested here
    detail::CheckDeviceCopyable< KernelType >();
    ^
types.cpp:64:9: note: in instantiation of function template specialization 'sycl::handler::single_task< XYZ,
      (lambda at types.cpp:64:32) >' requested here
    CGH.single_task< class XYZ >([=]() {

Reply:
Thanks.

Reply:
Fixed (updated the error message) in b24fab2650813f4790c50649d8446eac90eb494c